### PR TITLE
 Add Whitelisting Features to allow some external conntections

### PR DESCRIPTION
**Pull request by sva-mk**
Dear MISP-Guard maintainers,

I would like to submit a pull request addressing the need to allow external MISP feed connections while maintaining MISP-Guard's security posture.

### Motivation

MISP-Guard currently blocks all connections to and from unknown sources by default, which is an important security measure. However, many users rely on external MISP feeds to improve their threat intelligence. Therefore, it should be possible to whitelist URLs and domains so that feeds and MISPGuard can be used at the same time.

### Proposed Changes

I have added a new feature to MISP-Guard that allows users to define a whitelist of trusted external MISP feed URLs and domains. Any connection request from a server configured in `instances_host_mapping` to one of the entries in this whitelist will be allowed, while connections from unknown sources will still be blocked.

#### The changes include:

  1. Configuration Option: Added a new configuration option in the config.schema.json file to specify a list of trusted external `urls` and `domains`.  With the help of json arrays, users can add their MISP-Feeds to this list for whitelisting.
  2. Enhanced testing: The test suite has been extended to thoroughly cover the new whitelisting feature. Extended `pytest` suite to ensure correctness and security of changes.
  3. Documentation: Updated the project documentation to explain how to configure and use the new whitelisting feature effectively.

### Conclusion

I believe this feature will enhance the usability of MISP-Guard for users who rely on external MISP feeds. It strikes a balance between security and flexibility, allowing connections to known and trusted sources while maintaining the default security level.

I look forward to your feedback and am open to making any necessary adjustments based on your review. Thanks for considering this pull request.

Sincerely,
sva-mk

### CODE REVIEW

**Comment by righel** [comment]
Hello @sva-mk ,
The code and tests look good to me. Could you please change the term `whitelisting` to `allowlist`, both in the code and docs.
Thanks a lot for your contribution, indeed it would make the tool way more usable in real life scenarios.

File: src/config.schema.json [hunk 1]
```diff
@@ -2,6 +2,25 @@
     "$schema": "http://json-schema.org/draft-07/schema",
     "type": "object",
     "properties": {
+        "whitelisting": {
```
Comment (righel) [inline src/config.schema.json:5]: allowlist

File: src/mispguard.py [hunk 2]
```diff
@@ -128,32 +128,69 @@ def load(self, loader):
...
+                    logger.info("response from whitelisted url - skipping further processing")
+            except ForbiddenException as ex:
+                logger.error(ex)
+                return self.forbidden(flow, str(ex))
+            except Exception as ex:
+                logger.error(ex)
+                return self.forbidden(flow, "unexpected error, rejecting response")
+            
+    
+    def url_is_whitelisted(self, flow: http.HTTPFlow) -> bool:
```
Comment (righel) [inline src/mispguard.py:182]: url_is_allowed

File: src/mispguard.py [hunk 3]
```diff
@@ -128,32 +128,69 @@ def load(self, loader):
...
+        if flow.request.url in self.config["whitelisting"]["urls"]:
+            return True
+        else:
+            return False
+    
+    def domain_is_whitelisted(self, flow: http.HTTPFlow) -> bool:
```
Comment (righel) [inline src/mispguard.py:188]: domain_is_allowed

File: README.md [hunk 4]
```diff
@@ -94,6 +94,12 @@ sequenceDiagram
 * `blocked_attribute_categories`: Blocks if the event contains an attribute matching one of this categories.
 * `blocked_object_types`: Blocks if the event contains an object matching one of this types.
 
+**Whitelisting**
```
[the remark this answers is not in the document]
Comment (sva-mk) [reply README.md]: Done :)

File: src/config.schema.json (hunk 1 above)
Comment (sva-mk) [reply src/config.schema.json]: Done

File: src/mispguard.py (hunk 2 above)
Comment (sva-mk) [reply src/mispguard.py]: Done

File: src/mispguard.py (hunk 3 above)
Comment (sva-mk) [reply src/mispguard.py]: Done

**Comment by sva-mk** [comment]
Thank you for your feedback and positive assessment of the code and tests! I appreciate your suggestion to use the term "allowlist" instead of "whitelist" in the code and documentation.

**Pull request merged**